src/main/java/io/jenkins/plugins/artifact_manager_jclouds/BlobStoreProvider.java

```diff
@@ -69,6 +70,12 @@
     @NonNull
     public abstract BlobStoreContext getContext() throws IOException;
```

Comment:
This could perhaps be made non`abstract` (throwing an `AbstractMethodError` at runtime if neither method is overridden, to avoid a `StackOverflowError`), since in the other direction there is a `BlobStore.getContext()` (used in this plugin only to `close` the context for tests…which is likely a bug).

Reply:
I don't see where `BlobStore.getContext()` is used. In the test `BlobStoreProvider.getContext()` is used and on this context the `close()` is called.
I think it should stay abstract.

Reply:
> I don't see where `BlobStore.getContext()` is used.

Sorry, to clarify: I _think_ we could write something along the lines of:

```java
public /* final */ BlobStoreContext getContext() throws IOException {
    if (Util.isOverridden(BlobStoreProvider.class, getClass(), "getBlobStore")) {
        return getBlobStore().getContext();
    } else {
        throw new AbstractMethodError("override getBlobStore");
    }
}
public /* abstract */ BlobStore getBlobStore() throws IOException {
    return getContext().getBlobStore();
}
```

where the modifiers could be uncommented and the code simplified if we did not care to retain backward compatibility for hypothetical external implementations ([none are known](https://jenkins.io/doc/developer/extensions/artifact-manager-s3/#blobstoreprovider)).

Reply:
Actually `getContext()` returns a BlobStoreContext from jclouds.
So our 'getBlobStore()'  method is not called when you call `getContext().getBlobStore()`
We might consider getting rid of the `getContext()` method at all and use it only an a private method where required.

Reply:
ok, this might work as well, when speaking of not retaining backward compatibility one could also consider removing the `getContext()` method completely.
The getBlobStore should always be abstract I think.

Reply:
> when speaking of not retaining backward compatibility one could also consider removing the `getContext()` method completely

Yes, you could just inline it in that case.

> The getBlobStore should always be abstract

Then calls to this method against a (hypothetical) existing external implementation, which overrode/implemented `getContext`, would fail with an `AbstractMethodError`.

The above paired method idiom is common in Jenkins sources, by the way, when we change our minds about what methods an implementation of some interface (or abstract class) should define. Somewhat cleaner is to just deprecate the whole “interface” and introduce, say, a `BlobStoreProvider2` while having callers (anything enumerating this `ExtensionList`) convert from the old implementation.

Again, any of these tricks may just be overkill—the only other implementation of [JEP-202](https://jenkins.io/jep/202) I _know_ about is `azure-artifact-manager`, which uses the Azure APIs directly rather than via this plugin and jclouds.